src/security/context.cpp

```diff
@@ -643,16 +694,197 @@ class PolFinalWafCtx : public PolTaskCtx<PolFinalWafCtx> {
+                      "data to downstream filters immediately");
+        goto pass_downstream;
+      }
+    } else {  // !run_waf; we need more data
+      if (buffer_chain(*request.pool, in, true) != NGX_OK) {
+        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+    }
+  } else if (st == stage::AFTER_ON_REQ_WAF ||
+             st == stage::AFTER_ON_REQ_WAF_BLOCK) {
+    if (filter_ctx_.out) {  // first call after WAF ended
+      ngx_log_debug1(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                     "first filter call after WAF ended, req refcount=%d",
+                     request.main->count);
+      if (buffer_chain(*request.pool, in, false) != NGX_OK) {
+        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+
+    pass_downstream:
+      // pass saved buffers downstream
+      auto rc = ngx_http_next_request_body_filter(&request, filter_ctx_.out);
+
+      for (auto *cl = filter_ctx_.out; cl;) {
+        auto *ln = cl;
+        cl = cl->next;
+        ngx_free_chain(request.pool, ln);
+      }
+      filter_ctx_.out = nullptr;
+      filter_ctx_.out_last = &filter_ctx_.out;
+
+      return rc;
+    }
+
+    return ngx_http_next_request_body_filter(&request, in);
+  } else if (st == stage::SUSPENDED_ON_REQ_WAF) {
+    if (in) {
+      ngx_log_error(
+          NGX_LOG_NOTICE, request.connection->log, 0,
+          "unexpected filter call with data in stage SUSPENDED_ON_REQ_WAF");
+      // we're in a suspended state, so we don't expect to be called
+      // if we are called, it's a bit troubling, because the write that happens
+      // in the buffered chain in the next statement is not synchronized with
+      // the read that happens in the WAF thread.
+      if (buffer_chain(*request.pool, in, false) != NGX_OK) {
+        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+    }
+  } else {
+    ngx_log_error(NGX_LOG_NOTICE, request.connection->log, 0,
+                  "unexpected filter call in stage %d", st);
+    return ngx_http_next_request_body_filter(&request, in);
+  }
+
+  // we continue to call the filter chain, but don't pass any data
+  return ngx_http_next_request_body_filter(&request, nullptr);
+}
+
+ngx_int_t Context::buffer_chain(ngx_pool_t &pool, ngx_chain_t *in,
+                                bool consume) {
+  for (auto *in_ch = in; in_ch; in_ch = in_ch->next) {
+    ngx_chain_t *new_ch = ngx_alloc_chain_link(&pool);  // uninitialized
+    if (!new_ch) {
+      return NGX_ERROR;
+    }
+
+    auto *buf = in_ch->buf;
+    auto size = buf->last - buf->pos;
+    if (consume) {  // copy the buffer and consume the original
+      ngx_buf_t *new_buf = ngx_create_temp_buf(&pool, size);
+      if (!new_buf) {
+        return NGX_ERROR;
+      }
+
+      new_buf->last = ngx_copy(new_buf->pos, buf->pos, size);
+      buf->pos = buf->last;
+      new_buf->last_buf = buf->last_buf;
+      new_buf->tag = buf->tag;
+
+      new_ch->buf = new_buf;
+    } else {
+      new_ch->buf = buf;
```

Comment:
The only situation where we don't consume the buffer (consume: false) is on the first call after the WAF is completed, and we immediately pass a chain with this unconsumed buffer downstream, followed by  releasing the whole chain. So in effect, as far as this buffer is concerned, it's as if we just directly passed downstream the chain gotten by the filter, unaltered (actually we join the in chain with the copy of all the other buffers we got before calling the WAF, but that doesn't matter).

Now, there is one extra situation where we don't consume the buffer: if we're called while the WAF is suspended: but if this happens, it's a bug. As to your point about buffer reuse: by definition, a buffer that's not consumed (where pos != last) is a busy buffer, so it cannot be moved to the freelist.

Reply:
The **current** usage might be correct, but the implementation is not. While everything is fresh in your mind now and you understand all the interactions, consider the future. Think about the future you or a new contributor, who might need to modify this part of the module. Let's ensure it's as clear and correct as possible for them.

Reply:
Like I said in the last paragraph, buffer reuse is not a problem: if the buffer is busy it can't be reused (`ngx_chain_update_chains` moves buffers to free only if they're not busy). More relevant is your other point about recycled buffers in the other comment; those must be consumed. I'm not sure if those are ever produced by a request body filter in an nginx module, but it's indeed better be safe.